projects/extension/ai/semantic_catalog.py

```diff
@@ -0,0 +1,383 @@
+import json
+from collections.abc import Generator
+from typing import TypedDict
+
+
+class GeneratedDescription(TypedDict):
+    name: str
+    description: str
+
+
+def render_obj_sample(plpy, relation: str) -> str:
+    result = plpy.execute(f"select * from {relation}", 5)
+
+    ret_obj = f"""<data id="{relation}">"""
+    for r in result:
+        ret_obj += f"""\n  insert into {relation} ({', '.join(r.keys())}) values ({', '.join([str(v) for v in r.values()])});"""
+    ret_obj += "\n</data>"
+
+    return ret_obj
+
+
+def get_parsed_config(plpy, catalog_name: str, config: dict | None) -> dict:
+    if config is None:
+        result = plpy.execute(
+            f"select x.text_to_sql from ai.semantic_catalog x where x.catalog_name = '{catalog_name}';"
+        )
+        if len(result) > 0:
+            config = result[0]["text_to_sql"]
+
+    if config is None:
+        raise Exception("No config found")
+
+    return json.loads(config)  # type: dict
+
+
+def get_obj_description(
+    plpy, relation: str | None = None, fn: str | None = None
+) -> str:
+    objid = (
+        f"'{relation}'::pg_catalog.regclass::pg_catalog.oid"
+        if relation
+        else f"'{fn}'::regprocedure::pg_catalog.oid"
+    )
+    result = plpy.execute(
+        f"select ai.render_semantic_catalog_obj(0, 'pg_catalog.pg_class'::pg_catalog.regclass::pg_catalog.oid, {objid}) as description;"
+    )
```

Comment:
It might be better to have two separate queries. One for functions/procs. One for tables/views. Let's use parameterized queries instead of format strings for safety.

Reply:
Done in 4967a3d51918ff09a597e0f28a1e6daa6a6826a6